packages/babel-helper-modules/src/index.js

```diff
@@ -0,0 +1,302 @@
+import * as t from "babel-types";
+import template from "babel-template";
+import chunk from "lodash/chunk";
+
+import rewriteThis from "./rewrite-this";
+import rewriteLiveReferences from "./rewrite-live-references";
+import normalizeAndLoadModuleMetadata, {
+  hasExports,
+  isAnonymousImport,
+} from "./normalize-and-load-metadata";
+
+export { hasExports };
+
+/**
+ * Perform all of the generic ES6 module rewriting needed to handle initial
+ * module processing. This function will rewrite the majority of the given
+ * program to reference the modules described by the returned metadata,
+ * and returns a list of statements for use when initializing the module.
+ */
+export function rewriteModuleStatementsAndPrepareHeader(
+  path: NodePath,
+  { exportName, strict, allowTopLevelThis, strictMode, loose, noInterop },
+) {
+  const meta = normalizeAndLoadModuleMetadata(path, exportName, {
+    strict,
+    noInterop,
+  });
+
+  if (!allowTopLevelThis) {
+    rewriteThis(path);
+  }
+
+  rewriteLiveReferences(path, meta);
+
+  if (strictMode !== false && strict !== false) {
+    const hasStrict = path.node.directives.some(directive => {
+      return directive.value.value === "use strict";
+    });
+    if (!hasStrict) {
+      path.unshiftContainer(
+        "directives",
+        t.directive(t.directiveLiteral("use strict")),
+      );
+    }
+  }
+
+  const headers = [];
+  if (hasExports(meta) && !strict) {
+    headers.push(buildESModuleHeader(meta, loose /* enumerable */));
+  }
+
+  // Create all of the statically known named exports.
+  headers.push(...buildExportInitializationStatements(path, meta));
+
+  return { meta, headers };
+}
+
+/**
+ * Break down the module metadata into a simple array that contains the
+ * fields generally needed for compiling ES6 module support.
+ */
+export function getSourceMetadataArray(meta: ModuleMetadata) {
+  const lastNonAnonymousBlock = Array.from(
+    meta.source,
+  ).reduceRight((acc, [source, metadata]) => {
+    if (acc !== null) return acc;
+
+    if (isAnonymousImport(metadata)) return null;
+    return source;
+  }, null);
+
+  let inAnonymousBlock = lastNonAnonymousBlock === null;
+
+  const items = [];
+  for (const [source, metadata] of meta.source) {
+    const isAnonymous = isAnonymousImport(metadata);
+
+    items.push([source, metadata, isAnonymous, inAnonymousBlock]);
+
+    if (source === lastNonAnonymousBlock) inAnonymousBlock = true;
+  }
+
+  return items;
+}
+
+/**
+ * Flag a set of statements as hoisted above all else so that module init
+ * statements all run before user code.
+ */
+export function ensureStatementsHoisted(statements) {
```

Comment:
One day I'd like to be able to put all the init statements in a single `BlockStatement` and only hoist that, but the problem is how to expose the imports since they'd create `let` bindings in spec mode 🤔

Reply:
I'm on the fence. That's one option, but another would be to just put them all at the top of the file, and wrap everything that needs to be hoisted in a function declaration that will lazy-init the value to be hoisted. I haven't really settled on a solution either.